File: packages/python/python-shared/src/python_shared/http/resilience/tenant_rate_limiter.py
import time
import threading
from typing import Dict, Tuple

class TenantRateLimiter:
    def __init__(self, capacity: float = 100.0, fill_rate_per_sec: float = 50.0):
        self.capacity = capacity
        self.fill_rate_per_sec = fill_rate_per_sec
        self._buckets: Dict[str, Tuple[float, float]] = {}
        self._lock = threading.RLock()

    def allow_request(self, tenant_id: str, tokens: float = 1.0) -> bool:
        now = time.time()
        with self._lock:
            current_tokens, last_refill = self._buckets.get(tenant_id, (self.capacity, now))
            elapsed = now - last_refill
            refilled = min(self.capacity, current_tokens + elapsed * self.fill_rate_per_sec)

            if refilled >= tokens:
                self._buckets[tenant_id] = (refilled - tokens, now)
                return True
            self._buckets[tenant_id] = (refilled, now)
            return False

    def allow(self, key: str, tokens: float = 1.0) -> bool:
        return self.allow_request(key, tokens)

TokenBucketLimiter = TenantRateLimiter

File: packages/llm-obs-node-packages/shared-infra/src/http/resilience/tenant-rate-limiter.ts
/**
 * @file tenant-rate-limiter.ts
 * @description Outbound Per-Tenant Token Bucket Rate Limiter.
 * 
 * ALGORITHM & SPECIFICATION:
 * 1. Token Bucket Mathematical Model:
 *    - Each `tenantId` maintains a bucket with `capacity` (max tokens) and `fillRate` (tokens/sec).
 *    - On evaluation, calculates elapsed time $\Delta t = \text{now} - \text{lastRefill}$.
 *    - Refills tokens: $\text{tokens} = \min(\text{capacity}, \text{tokens} + \Delta t \times \text{fillRate})$.
 * 2. Admission Control:
 *    - `allowRequest(tenantId)`: Consumes 1 token if $\text{tokens} \ge 1$ and returns `true`.
 *    - Returns `false` if insufficient tokens remain.
 */

interface Bucket {
  tokens: number;
  lastRefill: number;
}

export class TenantRateLimiter {
  private readonly buckets = new Map<string, Bucket>();

  constructor(
    private readonly capacity = 100,
    private readonly fillRatePerSec = 50
  ) {}

  public allowRequest(tenantId: string): boolean {
    const now = Date.now();
    let bucket = this.buckets.get(tenantId);

    if (!bucket) {
      bucket = { tokens: this.capacity, lastRefill: now };
      this.buckets.set(tenantId, bucket);
    }

    const elapsedSec = (now - bucket.lastRefill) / 1000;
    bucket.tokens = Math.min(this.capacity, bucket.tokens + elapsedSec * this.fillRatePerSec);
    bucket.lastRefill = now;

    if (bucket.tokens >= 1) {
      bucket.tokens -= 1;
      return true;
    }

    return false;
  }
}
